marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/artifacts/core.sjs

```diff
@@ -56,11 +58,42 @@ function getArtifacts(artifactType) {
         queriesForAnd.push(cts.collectionQuery(coll));
     }
     if (queriesForAnd.length) {
-        const results = cts.search(queriesForAnd.length > 1 ? cts.andQuery(queriesForAnd) : queriesForAnd[0]);
-        for (const result of results) {
-            artifacts.push(result.toObject());
+        if (artifactType == "loadData") {
+            const results = cts.search(queriesForAnd.length > 1 ? cts.andQuery(queriesForAnd) : queriesForAnd[0]);
+            for (const result of results) {
+                artifacts.push(result.toObject());
+            }
+            return artifacts;
+        } else {
+            return getArtifactsGroupbyEntity(artifactType, queriesForAnd)
         }
     }
+    return [];
+}
+
+function getArtifactsGroupbyEntity(artifactType, queriesForAnd) {
+    const artifacts = [];
+    const entityNames = getEntityNames();
+    for (const ename of entityNames) {
```

Comment:
This is doing N queries, where N = entityNames.length. It's faster/better to do a single query - just pass entityNames into the cts.jsonPropertyValueQuery. Then iterate over the results with a map function. Can get it all done this simply (I often refer to https://www.w3schools.com/jsref/jsref_obj_array.asp for things like this):

```
function getArtifactsGroupByEntity(artifactType, queriesForAnd) {
  cts.search(cts.andQuery(queriesForAnd.concat(cts.jsonPropertyValueQuery("entityName", getEntityNames())).map(config => {
    const configByEntity = {
      name: config.path.to.entityName (I don't know the path)
    };
    switch(artifactType) {
    ....
    }
    return configByEntity;
  });
}
```

Note the use of "concat" to avoid modifying the queriesForAnd array that's passed in.

Reply:
`let entities = cts.search(cts.collectionQuery("http://marklogic.com/entity-services/models"))
const entityNames = entities.toArray().map(e => e.xpath("//info//title"));
const res = cts.search(cts.andQuery(queriesForAnd.concat(cts.jsonPropertyValueQuery("entityName", entityNames)))).toArray();

let res2 = res
    .reduce((acc, mapping) => {
      acc[mapping.entityName] = acc[mapping.entityName] || {};
      acc[mapping.entityName].mapping = (acc[mapping.entityName].mappings || []);
      acc[mapping.entityName].mapping.push(mapping);
      return acc;
    }, {});`

I tried to use the above code, and did not work.

Reply:
Fixed.